File: apps/oracle-service/scripts/deploy-market.ts
/**
 * HIP-3 Market Deployment Script
 * 
 * Deploys a HIP-3 perpetual market on Hyperliquid testnet.
 * 
 * Usage:
 *   NETWORK=testnet ts-node scripts/deploy-market.ts
 * 
 * Environment variables required:
 *   - HL_TESTNET_URL
 *   - HL_API_KEY
 *   - HL_API_SECRET
 *   - HL_DEX_NAME (optional, defaults to "WARMARKET")
 */

import crypto from 'crypto';
import { request } from 'undici';
import * as dotenv from 'dotenv';
import * as path from 'path';

// Load environment
const envFile = process.env.ENV_FILE ?? '.env.testnet';
dotenv.config({ path: path.resolve(process.cwd(), envFile) });

const HL_URL = process.env.HL_TESTNET_URL || 'https://api.hyperliquid-testnet.xyz';
const HL_API_KEY = process.env.HL_API_KEY;
const HL_API_SECRET = process.env.HL_API_SECRET;
const HL_DEX_NAME = process.env.HL_DEX_NAME || 'WARMARKET';

if (!HL_API_KEY || !HL_API_SECRET) {
  throw new Error('Missing HL_API_KEY or HL_API_SECRET');
}

/**
 * Register Asset Action
 * 
 * This is the first step in deploying a HIP-3 market.
 * Registers the asset with market parameters.
 */
interface RegisterAsset {
  type: 'registerAsset';
  name: string;  // Asset name (e.g., "GOLD-TEST")
  szDecimals: number;  // Size decimals (recommended: 0-2 for $1-10 increments)
  maxLeverage?: number;  // Maximum leverage (default: 20)
  oracle?: {
    // Oracle definition - points to your oracle service
    // This may need to be your API wallet address or a contract address
    address?: string;
  };
}

/**
 * Deploy Perpetual Market
 * 
 * Sends the registerAsset action to deploy a new HIP-3 market.
 */
async function deployMarket(config: {
  assetName: string;
  szDecimals: number;
  maxLeverage?: number;
}): Promise<void> {
  const { assetName, szDecimals, maxLeverage = 20 } = config;

  console.log(`\n🚀 Deploying HIP-3 market: ${assetName}`);
  console.log(`   DEX: ${HL_DEX_NAME}`);
  console.log(`   Size Decimals: ${szDecimals}`);
  console.log(`   Max Leverage: ${maxLeverage}x\n`);

  const registerAsset: RegisterAsset = {
    type: 'registerAsset',
    name: assetName,
    szDecimals,
    maxLeverage,
  };

  const action = {
    type: 'perpDeploy',
    registerAsset,
  };

  const payload = {
    action,
  };

  const body = JSON.stringify(payload);
  
  // Sign the request (authentication method may vary - verify with HL docs)
  const signature = crypto.createHmac('sha256', HL_API_SECRET)
    .update(body)
    .digest('hex');

  const endpoint = `${HL_URL}/exchange`;
  
  console.log(`📡 Sending to: ${endpoint}`);
  console.log(`📦 Payload:`, JSON.stringify(payload, null, 2));
  // ⚠️ SECURITY: Never log HL_API_SECRET or full API keys in production

  try {
    const response = await request(endpoint, {
      method: 'POST',
      body,
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${HL_API_KEY}:${signature}`,
        'X-API-Key': HL_API_KEY,
      },
    });

    const responseText = await response.body.text();
    
    if (response.statusCode >= 400) {
      console.error(`❌ Deployment failed (${response.statusCode}):`, responseText);
      throw new Error(`Deployment failed: ${responseText}`);
    }

    const result = JSON.parse(responseText);
    console.log(`✅ Deployment successful!`);
    console.log(`📋 Response:`, JSON.stringify(result, null, 2));
    
    console.log(`\n📝 Next steps:`);
    console.log(`   1. Update HL_MARKET_ID in .env.testnet: ${assetName}`);
    console.log(`   2. Start your oracle service: npm run dev`);
    console.log(`   3. Verify market appears on Hyperliquid testnet`);
    console.log(`   4. Test price publishing via setOracle\n`);

  } catch (error) {
    console.error(`❌ Error deploying market:`, error);
    throw error;
  }
}

/**
 * Main execution
 */
async function main() {
  // Example: Deploy GOLD-TEST market
  // Adjust these parameters based on your needs
  
  const marketConfig = {
    assetName: 'GOLD-TEST',  // Change to your desired asset name
    szDecimals: 2,  // For $0.01 increments (if price ~$2000)
    maxLeverage: 20,  // 20x leverage
  };

  // Uncomment to deploy:
  // await deployMarket(marketConfig);

  console.log(`
⚠️  DEPLOYMENT SCRIPT READY

This script is ready to deploy a HIP-3 market, but you should:

1. Verify the exact API structure with Hyperliquid docs
2. Check if testnet requires staking (may be waived)
3. Confirm authentication method (HMAC vs wallet signature)
4. Review market parameters before deploying

To deploy, uncomment the deployMarket() call above and run:
  NETWORK=testnet ts-node scripts/deploy-market.ts

Recommended first market: GOLD-TEST (solo asset, simple to test)
  `);
}

main().catch((error) => {
  console.error('Fatal error:', error);
  process.exit(1);
});

File: apps/oracle-service/scripts/deploy-market.py
#!/usr/bin/env python3
"""
HIP-3 Market Deployment Script (Python)
Deploys a HIP-3 perpetual market on Hyperliquid testnet using the official Python SDK.

Usage:
    NETWORK=testnet python3 scripts/deploy-market.py

Environment variables required:
    - HL_TESTNET_URL (default: https://api.hyperliquid-testnet.xyz)
    - HL_API_KEY (API wallet address)
    - HL_API_SECRET (API wallet private key)
"""

import os
import sys
from pathlib import Path
from dotenv import load_dotenv

# Load environment
env_file = os.getenv('ENV_FILE', '.env.testnet')
load_dotenv(env_file)

# Import Hyperliquid SDK
try:
    import eth_account
    from hyperliquid.exchange import Exchange
    from hyperliquid.utils import constants
except ImportError:
    print("❌ Error: Hyperliquid Python SDK not installed")
    print("Install with: pip3 install hyperliquid-python-sdk")
    sys.exit(1)

# Get environment variables
HL_API_KEY = os.getenv('HL_API_KEY')
HL_API_SECRET = os.getenv('HL_API_SECRET')
HL_TESTNET_URL = os.getenv('HL_TESTNET_URL', 'https://api.hyperliquid-testnet.xyz')

if not HL_API_KEY or not HL_API_SECRET:
    print("❌ Error: Missing HL_API_KEY or HL_API_SECRET")
    sys.exit(1)

def create_market():
    """Create a HIP-3 perpetual market on Hyperliquid testnet."""
    
    print("\n🚀 Creating HIP-3 market: XAU-TEST")
    print(f"   API URL: {HL_TESTNET_URL}")
    print(f"   API Wallet: {HL_API_KEY}\n")
    
    # Initialize wallet and exchange
    try:
        wallet = eth_account.Account.from_key(HL_API_SECRET)
        print(f"✅ Wallet initialized: {wallet.address}")
        
        # Use testnet URL
        exchange = Exchange(wallet, constants.TESTNET_API_URL)
        print(f"✅ Exchange initialized for testnet\n")
    except Exception as e:
        print(f"❌ Error initializing wallet/exchange: {e}")
        sys.exit(1)
    
    # Market configuration
    market_config = {
        'symbol': 'XAU-TEST',
        'szDecimals': 2,
        'initialOraclePrice': '1924.55',
        'marginTableId': 0,
        'onlyIsolated': False,
    }
    
    print("📋 Market Configuration:")
    print(f"   Symbol: {market_config['symbol']}")
    print(f"   Size Decimals: {market_config['szDecimals']}")
    print(f"   Initial Oracle Price: ${market_config['initialOraclePrice']}")
    print(f"   Margin Table ID: {market_config['marginTableId']}")
    print(f"   Isolated Only: {market_config['onlyIsolated']}\n")
    
    # Create the perpDeploy action
    # Based on Hyperliquid Python SDK structure
    action = {
        'type': 'perpDeploy',
        'registerAsset': {
            'type': 'registerAsset',
            'coin': market_config['symbol'],
            'szDecimals': market_config['szDecimals'],
            'oraclePx': market_config['initialOraclePrice'],
            'marginTableId': market_config['marginTableId'],
            'onlyIsolated': market_config['onlyIsolated'],
        }
    }
    
    print("📦 Action payload:")
    import json
    print(json.dumps(action, indent=2))
    print()
    
    # Use the exchange object to send the action
    # The Python SDK has a specific method: perp_deploy_register_asset
    try:
        print("📡 Sending createMarket request...\n")
        
        # Method signature: perp_deploy_register_asset(
        #   dex: str,
        #   max_gas: Optional[int],
        #   coin: str,
        #   sz_decimals: int,
        #   oracle_px: str,
        #   margin_table_id: int,
        #   only_isolated: bool,
        #   schema: Optional[PerpDexSchemaInput]
        # )
        
        dex_name = os.getenv('HL_DEX_NAME', 'WARMARKET')
        max_gas = None  # Optional, let Hyperliquid decide
        
        print("📤 Calling perp_deploy_register_asset:")
        print(f"   DEX: {dex_name}")
        print(f"   Coin: {market_config['symbol']}")
        print(f"   Size Decimals: {market_config['szDecimals']}")
        print(f"   Oracle Price: {market_config['initialOraclePrice']}")
        print(f"   Margin Table ID: {market_config['marginTableId']}")
        print(f"   Only Isolated: {market_config['onlyIsolated']}")
        print()
        
        result = exchange.perp_deploy_register_asset(
            dex=dex_name,
            max_gas=max_gas,
            coin=market_config['symbol'],
            sz_decimals=market_config['szDecimals'],
            oracle_px=market_config['initialOraclePrice'],
            margin_table_id=market_config['marginTableId'],
            only_isolated=market_config['onlyIsolated'],
            schema=None,  # Optional schema
        )
        
        print("✅ Market creation response:")
        print(json.dumps(result, indent=2))
        
        # Extract assetId from response
        if result.get('status') == 'ok':
            response_data = result.get('response', {})
            asset_id = response_data.get('assetId') or response_data.get('data', {}).get('assetId')
            
            if asset_id:
                print(f"\n🎯 Market created successfully!")
                print(f"   Asset ID: {asset_id}")
                print(f"   Symbol: {market_config['symbol']}")
                print(f"\n📝 Next steps:")
                print(f"   1. Update .env.testnet:")
                print(f"      HL_MARKET_ID={asset_id}")
                print(f"      HL_MARKET_SYMBOL={market_config['symbol']}")
                print(f"      HL_PUBLISH_ENABLED=true")
                print(f"   2. Start oracle service: npm run dev")
                print(f"   3. Verify at: https://app.hyperliquid-testnet.xyz/perps\n")
                return asset_id
            else:
                print("\n⚠️  Warning: Market created but assetId not found in response")
                print("   Check the response above for the asset ID")
        else:
            print(f"\n❌ Market creation failed: {result}")
            
    except Exception as e:
        print(f"\n❌ Error creating market: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)

if __name__ == '__main__':
    create_market()

